### PR TITLE
Automatically bridge eligible dependencies on pulumi convert from terraform

### PR DESCRIPTION
**Pull request by brandonpollack23**
The culmination! With all the additions to package blocks,
paramaterization, loaders, etc, we can now finally automatically
generate and link SDKs on calls to `pulumi convert` for non pulumified
sdks!

This change does a few things:

* Generate sdks for parameterized terraform providers on convert and
  link them to the converted project.
* Move package loader to use V2 and package descriptor so the schema
  loading works correctly by parameterizing the provider.
* Explicilty check if SupportPack is set in order to determine if the
  packages are flat are not (support pack doesn't support them, so it
  uses the correct generated sdk structure).

Fixes pulumi/pulumi#18019

### CODE REVIEW

File: pkg/cmd/pulumi/convert.go
Comment (brandonpollack23) [reply pkg/cmd/pulumi/convert.go:333]: agreed.

File: pkg/cmd/pulumi/convert.go
Comment (brandonpollack23) [reply pkg/cmd/pulumi/convert.go:375]: I specifically here am just checking if we already generated pcl using terraform provdier so I can print warnings to the user and have diags for debugging if stuff goes wrong.

I think I shouldn't exit though, since we still need to install the dynamic bridge if it it is not present.

File: pkg/cmd/pulumi/packagecmd/package.go
Comment (brandonpollack23) [inline pkg/cmd/pulumi/packagecmd/package.go:1]: just a head's up, this was moved to make sharing eaiser with other packages

File: pkg/codegen/convert/plugin_mapper.go
Comment (Frassle) [inline pkg/codegen/convert/plugin_mapper.go:124]: Why are we renaming this? It's only _ever_ going to be a pulumi provider.

File: pkg/codegen/schema/loader_server.go
Comment (Frassle) [inline pkg/codegen/schema/loader_server.go:55]: Surprised nothing needed this before. This is small enough and obvious enough we could merge this change as is if it was its own PR.

File: pkg/cmd/pulumi/convert.go
Comment (Frassle) [inline pkg/cmd/pulumi/convert.go:612]:
```suggestion
			return fmt.Errorf("creating package schema: %w", err)
```
Typo but also this applies to a lot of the error messages in this file, do not prefix them all with error: https://preslav.me/2023/04/14/golang-error-handling-is-a-form-of-storytelling/#crafting-a-good-message-is-key

File: pkg/cmd/pulumi/convert.go
Comment (Frassle) [reply pkg/cmd/pulumi/convert.go:372]: I think this whole block should go. The warning about no parameterisation is a warning for the terraform provider to decided on, and we don't need to warn the user about every local sdk we build. They should just happen and be expected.

File: pkg/cmd/pulumi/convert.go
Comment (Frassle) [inline pkg/cmd/pulumi/convert.go:392]:
```suggestion
			pCtx.Diag.Warningf(diag.Message("", "failed to install provider %q: %v"), provider, err)
```
We only ever install pulumi providers, we don't need to distinguish here.

File: pkg/codegen/convert/plugin_mapper.go
Comment (brandonpollack23) [reply pkg/codegen/convert/plugin_mapper.go:124]: that is true but I wanted to make it explicit (especially when I was going to warn to let a user/us know an SDK was being generated for debugging purposes) that this was a pulumi provider and does not handle TF bridging.

File: pkg/codegen/schema/loader_server.go
Comment (brandonpollack23) [reply pkg/codegen/schema/loader_server.go:55]: It is in it's own pr now to be submit separately #18021

**Comment by brandonpollack23** [comment]
I believe this is blocked on a pulumi/pulumi-terraform-bridge release

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:52]: I don't think this belongs in `cmd/pulumi/cmd`. My sense is it should live in `cmd/pulumi/packagecmd`, but be publicly exported for `convert` to use. `cmd` is more for things that affect all commands. This would also reduce other diffs where you've had to rename to `pcmd`.

Side note -- I think this could be `GeneratePackage`, and the linking `LinkPackage`

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:181]: As mentioned above, perhaps this could be `LinkPackage`? That'd track with the private helpers it calls. Its doc comment probably needs changing in light of the behaviour updates too.

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:203]: This comment is probably outdated now. Same for any of the other `link*` helpers that are now automating stuff.

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:229]: Debug noise?

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:233]:
```suggestion
		// Assume npm if no package manager is specified
```

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:332]:
```suggestion
		// Assume pip if no package manager is specified
```

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:360]: Feel like this could be worded differently. One directory deeper? Equal was not what I was expecting. Might just be me though.

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:364]: What's pathPrefix here, sorry?

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:376]: Does `filepath.Join` not take care of this kind of thing?

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:371]: And if there is parameterization, we do nothing? I feel like maybe the original comment above could cover this if we put a couple of examples in it, maybe?

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:440]: Feels like we want some test cases for this.

File: pkg/cmd/pulumi/cmd/project_sdk_link.go
Comment (lunaris) [inline pkg/cmd/pulumi/cmd/project_sdk_link.go:571]: I don't think this is worth sharing and it's bloating this PR -- let's put it back.

File: pkg/cmd/pulumi/convert/convert_test.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert_test.go:56]: I think this is a bad conflict merge -- you want `testdata` on your side here I think since that got moved under `convert/`.

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:497]: Maybe make this `ParseDirectory`? Would fit nicely into `BindDirectory` and save you duplicating the `ReadDir` code here and in the `pcl` package.

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:499]:
```suggestion
		return nil, nil, fmt.Errorf("could not parse PCL files: %w", err)
```

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:493]: Maybe this is more helpful for debugging?
```suggestion
		return nil, nil, fmt.Errorf("could not read %s: %w", sourceDirectory, err)
```

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:490]: Dead TODO?

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:515]: Wonder if this is worth pulling out into `pcl.ReadAllPackageDescriptors` or similar. Or maybe rename the original to `ReadFilePackageDescriptors` and introduce `ReadFilesPackageDescriptors`? Would de-dupe this with `pcl` and make this smaller again.

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:553]: Perhaps shave this to:
```go
if pkg.Parameterization == nil {
  continue
}

parameterizationValue := pkg.Parameterization.Value
```
(or just use `pkg.Parameterization.Value` directly thereafter maybe)

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:571]: Yea feels like we can leave out the `parameterizationValue` variable

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:616]: Is this `return` necessary?

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:624]: Looks like there's only one use of this -- kill it? Think that means `diagf` can go too.

File: pkg/cmd/pulumi/convert/convert.go
Comment (lunaris) [inline pkg/cmd/pulumi/convert/convert.go:585]: I think for comments like this, it can be valuable to explain how things would misbehave if we didn't do it -- as it stands this leaves me with more questions than answers.

**Review by tgummerer — APPROVED** [review]
I have a few suggestions inline that are probably worth addressing before merging.

**Pull request merged**
